src/realm/query_expression.hpp

```diff
@@ -3117,11 +3097,16 @@ class ListColumnAggregate : public Subexpr2<typename Operation::ResultType> {
             auto list_ref = to_ref(list_refs[i].get_int());
             Operation op;
             if (list_ref) {
-                BPlusTree<StorageType> list(alloc);
-                list.init_from_ref(list_ref);
-                size_t s = list.size();
-                for (unsigned j = 0; j < s; j++) {
-                    op.accumulate(list.get(j));
+                if constexpr (realm::is_any_v<T, ObjectId, Int, Bool, UUID>) {
```

Comment:
Consider using `if constexpr (!ObjectTypeTraits<T>::is_self_contained_null)` for maintainability. It'll be slower on float/double because of the extra check but the current (and past) code only works because we fake util::Optional<Float> as a special NaN internally. I guess I could argue either way because of the performance cost vs misusing the API by constructing a BPTree<Float> on a nullable float array which happens to work because of the implementation. So I guess I have no strong opinions because this just brings up the nullability discussion which is not part of this PR, but I thought I should say something because it makes me slightly uncomfortable.

Reply:
I like the idea of consolidating the knowledge of what types should be wrapped in Optional to one place, but changing this to check `self_contained_null` has some weird effects. Float and double actually need it to be true (`table.columns<util::Optional<float>>()` currently doesn't work) as do things like util::Optional<int64_t>, so it seems like it's a bigger refactoring than just changing this one spot.

Reply:
Alright, thanks for checking. This section of code really just spotlights the nullability design inconsistency which we don't have a clean solution for yet, so whatever works here is fine by me.